Store/StoreBraintreePaymentProvider.php

```diff
@@ -0,0 +1,455 @@
+	 * @param string $card_verification_value optional. Card verification value
+	 *                                         used for fraud prevention.
+	 *
+	 * @return StorePaymentMethodTransaction the transaction object for the
+	 *                                        payment. This object contains the
+	 *                                        transaction date and identifier.
+	 *
+	 * @sensitive $card_number
+	 * @sensitive $card_verification_value
+	 */
+	public function pay(StoreOrder $order, $card_number,
+		$card_verification_value = null)
+	{
+		$request = array(
+			'amount' => $this->formatCurrency($order->total),
+			'orderId' => $this->getOrderId($order),
+			'creditCard' => $this->getCreditCard(
+				$order,
+				$card_number,
+				$card_verification_value
+			),
+			'options' => array(
+				'submitForSettlement' => true,
+			),
+		);
+
+		if ($order->billing_address instanceof StoreOrderAddress) {
+			$request['billing'] = $this->getBillingAddress(
+				$order->billing_address
+			);
+		}
+
+		if ($order->account instanceof StoreAccount) {
+			$request['customer'] = $this->getCustomer($order->account);
+		}
+
+		$custom_fields = $this->getCustomFields($order);
+		if (count($custom_fields) > 0) {
+			$request['customFields'] = $custom_fields;
+		}
+
+		// do transaction
+		$this->setConfig();
+		$response = Braintree_Transaction::sale($request);
+
+		// check for errors and throw exception
+		if (!$response->success) {
+			throw new StorePaymentBraintreeException(
+				'Error processing Braintree payment.',
+				0,
+				$response
+			);
+		}
+
+		$class_name = SwatDBClassMap::get('StorePaymentMethodTransaction');
+		$transaction = new $class_name();
+
+		$transaction->transaction_type = StorePaymentRequest::TYPE_PAY;
+		$transaction->transaction_id = $response->transaction->id;
+		$transaction->createdate = new SwatDate();
+		$transaction->createdate->toUTC();
+
+		return $transaction;
+	}
+
+	// }}}
+	// {{{ public function getExceptionMessageId()
+
+	public function getExceptionMessageId(Exception $e)
+	{
+		if ($e instanceof Braintree_Exception_Authentication ||
+			$e instanceof Braintree_Exception_Authorization ||
+			$e instanceof Braintree_Exception_Configuration ||
+			$e instanceof Braintree_Exception_ServerError ||
+			$e instanceof Braintree_Exception_UpgradeRequired) {
+			return 'payment-error';
+		}
+
+		if ($e instanceof StorePaymentBraintreeException) {
+			$response = $e->getResponse();
+
+			// data validation error(s)
+			if ($response->errors->deepSize() > 0) {
+				return 'payment-error';
+			}
+
+			$transaction = $response->transaction;
+
+			// transaction error
+			if ($transaction->status === 'processor_declined') {
+				switch ($transaction->processorResponseCode) {
+				case 2004: // expired card
+					return 'card-expired';
+				case 2010: // issuer declined CVV
+					return 'card-verification-value';
+				case 2060: // address verification failed
+					return 'address-mismatch';
```

Comment:
StoreCheckoutConfirmation doesn't have error handling for 'card-error' or for 'card-not-valid'. I'll add that and split out the error types.